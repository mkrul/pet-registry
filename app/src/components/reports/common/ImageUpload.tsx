import React from "react";
import { Button, FormControl, Alert } from "@mui/material";
import { CloudUpload } from "@mui/icons-material";
import { ImageUploadProps } from "../../../types/common/Image";
import { commonInputStyles } from "../../../styles/commonStyles";
import Tip from "../../common/Tip";

export const ImageUpload: React.FC<ImageUploadProps> = ({
  onImageSelect,
  onFileChange,
  preview,
  disabled,
  onImageLoad,
  onImageError,
  error
}) => {
  const handleImageChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files && e.target.files[0]) {
      const file = e.target.files[0];

      if (file.size > 5 * 1024 * 1024) {
        alert("File size exceeds 5MB.");
        return;
      }

      const validTypes = ["image/jpeg", "image/png", "image/gif"];
      if (!validTypes.includes(file.type)) {
        alert("Unsupported file type. Please upload a JPEG, PNG, or GIF image.");
        return;
      }

      const reader = new FileReader();
      reader.onloadend = () => {
        onImageSelect?.(file, reader.result as string);
        onFileChange?.(e);
      };
      reader.readAsDataURL(file);
    }
  };

  return (
    <div className="space-y-2">
      <label className="text-lg font-medium text-gray-900 mb-2">Photo:</label>
      <Tip>
        The best photos are taken outdoors with good lighting. Your photo should be well lit, with
        the animal's face centered in the frame.
      </Tip>
      <div className="mt-1">
        <div className="relative">
          <input
            type="file"
            name="image"
            accept="image/*"
            onChange={handleImageChange}
            style={{
              opacity: 0,
              position: "absolute",
              top: 0,
              left: 0,
              width: "100%",
              height: "100%",
              cursor: "pointer",
              zIndex: 1
            }}
            disabled={disabled}
            data-testid="file-input"
            required
          />
          <Button
            variant="outlined"
            startIcon={<CloudUpload />}
            disabled={disabled}
            sx={{
              ...commonInputStyles,
              width: "fit-content",
              position: "relative",
              borderColor: error ? "error.main" : undefined
            }}
          >
            Choose File
          </Button>
        </div>
        {error && (
          <Alert severity="error" sx={{ mb: 2 }}>
            {error}
          </Alert>
        )}
        {preview && (
          <div className="mt-2 relative w-48 h-48">
            <img
              src={preview}
              alt="Preview"
              className="object-cover w-full h-full rounded-md"
              onLoad={onImageLoad}
              onError={onImageError}
            />
          </div>
        )}
      </div>
    </div>
  );
};
