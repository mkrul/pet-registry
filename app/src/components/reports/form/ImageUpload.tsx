import React from "react";
import { FormControl, Button } from "@mui/material";
import { CloudUpload } from "@mui/icons-material";
import { ImageUploadProps } from "../../../types/common/Image";

export const ImageUpload: React.FC<ImageUploadProps> = ({ onImageSelect, preview, disabled }) => {
  const handleImageChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files && e.target.files[0]) {
      const file = e.target.files[0];

      // Validate file size (max 5MB)
      if (file.size > 5 * 1024 * 1024) {
        alert("File size exceeds 5MB.");
        return;
      }

      // Validate file type
      const validTypes = ["image/jpeg", "image/png", "image/gif"];
      if (!validTypes.includes(file.type)) {
        alert("Unsupported file type. Please upload a JPEG, PNG, or GIF image.");
        return;
      }

      onImageSelect(file);
    }
  };

  return (
    <div className="mt-8">
      <h2 className="text-lg font-medium text-gray-900 mb-2">Upload Image</h2>
      <FormControl fullWidth>
        <Button
          component="label"
          variant="outlined"
          startIcon={<CloudUpload />}
          sx={{ mt: 1, width: "fit-content" }}
          disabled={disabled}
        >
          Choose File
          <input
            type="file"
            name="image"
            accept="image/*"
            onChange={handleImageChange}
            style={{ display: "none" }}
            disabled={disabled}
            data-testid="file-input"
          />
        </Button>
        {preview && (
          <div className="relative w-32 h-32 mt-3">
            <img src={preview} alt="Selected" className="object-cover w-full h-full rounded-md" />
          </div>
        )}
      </FormControl>
    </div>
  );
};
