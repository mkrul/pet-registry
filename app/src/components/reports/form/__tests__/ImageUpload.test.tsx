import React from "react";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { vi, describe, it, expect, beforeEach, afterEach } from "vitest";
import { ImageUpload } from "../ImageUpload";

describe("ImageUpload", () => {
  const mockOnImageSelect = vi.fn();
  const defaultProps = {
    onImageSelect: mockOnImageSelect,
    preview: undefined,
    disabled: false
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  afterEach(() => {
    cleanup();
  });

  it("renders correctly", () => {
    const { container } = render(<ImageUpload {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("displays the choose file button", () => {
    render(<ImageUpload {...defaultProps} />);
    const button = screen.getByRole("button", { name: /choose file/i });
    expect(button).toBeDefined();
  });

  it("shows preview image when preview prop is provided", () => {
    const previewUrl = "test-image-url.jpg";
    render(<ImageUpload {...defaultProps} preview={previewUrl} />);
    const previewImage = screen.getByAltText("Selected");
    expect(previewImage).toBeDefined();
    expect(previewImage).toHaveAttribute("src", previewUrl);
  });

  it("handles file selection with valid image", () => {
    const file = new File(["test"], "test.jpg", { type: "image/jpeg" });
    render(<ImageUpload {...defaultProps} />);

    const input = screen.getByTestId("file-input") as HTMLInputElement;
    fireEvent.change(input, { target: { files: [file] } });

    expect(mockOnImageSelect).toHaveBeenCalledWith(file);
  });

  it("rejects files larger than 5MB", () => {
    const alertSpy = vi.spyOn(window, "alert").mockImplementation(() => {});
    // Create a mock file that's larger than 5MB
    const largeFile = new File([new ArrayBuffer(6 * 1024 * 1024)], "large.jpg", {
      type: "image/jpeg"
    });

    render(<ImageUpload {...defaultProps} />);

    const input = screen.getByTestId("file-input") as HTMLInputElement;
    fireEvent.change(input, { target: { files: [largeFile] } });

    expect(alertSpy).toHaveBeenCalledWith("File size exceeds 5MB.");
    expect(mockOnImageSelect).not.toHaveBeenCalled();

    alertSpy.mockRestore();
  });

  it("rejects invalid file types", () => {
    const alertSpy = vi.spyOn(window, "alert").mockImplementation(() => {});
    const invalidFile = new File(["test"], "test.txt", { type: "text/plain" });

    render(<ImageUpload {...defaultProps} />);

    const input = screen.getByTestId("file-input") as HTMLInputElement;
    fireEvent.change(input, { target: { files: [invalidFile] } });

    expect(alertSpy).toHaveBeenCalledWith(
      "Unsupported file type. Please upload a JPEG, PNG, or GIF image."
    );
    expect(mockOnImageSelect).not.toHaveBeenCalled();

    alertSpy.mockRestore();
  });

  it("disables input when disabled prop is true", () => {
    render(<ImageUpload {...defaultProps} disabled={true} />);

    const button = screen.getByRole("button", { name: /choose file/i });
    const input = screen.getByTestId("file-input") as HTMLInputElement;

    expect(button).toHaveClass("Mui-disabled");
    expect(input).toBeDisabled();
  });

  it("accepts valid image types", async () => {
    const validTypes = ["image/jpeg", "image/png", "image/gif"];

    for (const type of validTypes) {
      cleanup();
      const file = new File(["test"], `test.${type.split("/")[1]}`, { type });
      render(<ImageUpload {...defaultProps} />);

      const input = screen.getByTestId("file-input") as HTMLInputElement;
      fireEvent.change(input, { target: { files: [file] } });

      expect(mockOnImageSelect).toHaveBeenCalledWith(file);
    }
  });
});
