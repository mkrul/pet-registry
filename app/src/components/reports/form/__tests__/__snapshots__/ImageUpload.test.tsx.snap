// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageUpload > renders correctly 1`] = `
<div>
  <div
    class="mt-8"
  >
    <h2
      class="text-lg font-medium text-gray-900 mb-2"
    >
      Upload Image
    </h2>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-ytlejw-MuiFormControl-root"
    >
      <label
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-3545tq-MuiButtonBase-root-MuiButton-root"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1jsm74t-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="CloudUploadIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96M14 13v4h-4v-4H7l5-5 5 5z"
            />
          </svg>
        </span>
        Choose File
        <input
          accept="image/*"
          data-testid="file-input"
          name="image"
          style="display: none;"
          type="file"
        />
      </label>
    </div>
  </div>
</div>
`;
